=== modified file 'sale/sale.py'
--- sale/sale.py	2012-10-23 16:05:04 +0000
+++ sale/sale.py	2013-11-14 17:05:34 +0000
@@ -859,6 +859,7 @@
     def product_id_change(self, cr, uid, ids, pricelist, product, qty=0,
             uom=False, qty_uos=0, uos=False, name='', partner_id=False,
             lang=False, update_tax=True, date_order=False, packaging=False, fiscal_position=False, flag=False, context=None):
+        original_context = context.copy()
         context = context or {}
         lang = lang or context.get('lang',False)
         if not  partner_id:
@@ -943,11 +944,13 @@
                     'Please set one before choosing a product.')
             warning_msgs += _("No Pricelist ! : ") + warn_msg +"\n\n"
         else:
+            ctx = original_context.copy()
+            ctx.update({
+                'uom': uom or result.get('product_uom'),
+                'date': date_order,
+            })
             price = self.pool.get('product.pricelist').price_get(cr, uid, [pricelist],
-                    product, qty or 1.0, partner_id, {
-                        'uom': uom or result.get('product_uom'),
-                        'date': date_order,
-                        })[pricelist]
+                    product, qty or 1.0, partner_id, context=ctx)[pricelist]
             if price is False:
                 warn_msg = _("Cannot find a pricelist line matching this product and quantity.\n"
                         "You have to change either the product, the quantity or the pricelist.")

